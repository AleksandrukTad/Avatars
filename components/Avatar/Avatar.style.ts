import styled, { css } from "styled-components";

interface ICircleProps {
  active: boolean;
  borderColor: string | undefined;
  overlay?: string;
  loading: boolean;
  extended: boolean;
}

interface ITextProps {
  color?: string;
}

export const Wrapper = styled.div`
  display: flex;
  height: 40px;
  width: 40px;
  justify-content: center;
  align-items: center;

  & + & {
    margin-top: 1rem;
  }
`;

export const Circle = styled.div<ICircleProps>`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 40px;
  width: 40px;
  border-radius: 50%;
  background-color: rgb(135, 206, 250);

  ${({ active }) =>
    !active &&
    css`
      transition: 0.25s ease-in-out;

      ${(props: ICircleProps) => props.extended && "transform: scale(1.2)"};
  };
`};

  /* Border */
  &::after {
    position: absolute;
    content: "";
    height: 40px;
    width: 40px;
    border-radius: 50%;
    background-color: transparent;

    ${({ borderColor, loading, theme }) =>
      `border: solid 5px  ${loading ? theme.Loading_Color : borderColor}`};
  }
  /* Overlay */
  &::before {
    position: absolute;
    content: "";
    width: 100%;
    height: 100%;
    border-radius: 50%;
    ${({ overlay }) => overlay && `background-color: ${overlay}`};
    opacity: 0.5;
  }
`;

export const Text = styled.p<ITextProps>`
  color: ${({ color }) => color || "black"};
  margin: 0;
  padding: 0;
  font-weight: bold;
`;

export const Canvas = styled.canvas`
  position: absolute;
`;
