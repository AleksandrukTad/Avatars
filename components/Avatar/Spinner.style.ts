import styled, { keyframes } from "styled-components";

const spinnerSize = 40;

interface ISpinnerProps {
  loading: boolean;
}

interface ICircleProps {
  theme: {
    Loading_Background_Color: string;
  };
}

const spinnerAnimation = keyframes`{
  0% {
    stroke-dashoffset: 0
    transform: rotate(-90deg);
  } 25% {
    transform: rotate(0deg);
  } 50% {
    transform: rotate(0deg);
  }100% {
    transform: rotate(0deg);
    stroke-dashoffset: -125;
  }
}`;

export const Spinner = styled.svg<ISpinnerProps>`
  ${({ loading }) => (loading ? "display: block" : "display: none")};

  position: absolute;
  width: ${spinnerSize}px;
  height: ${spinnerSize}px;
  transform: scale(1.5);
  z-index: 99999;
`;

export const Circle = styled.circle<ICircleProps>`
  fill: transparent;
  transform: rotate(-90deg);
  stroke: ${({ theme }) => theme.Loading_Background_Color};
  stroke-width: 4;
  stroke-linecap: round;
  stroke-dasharray: ${3.14 * spinnerSize};
  transform-origin: ${0.5 * spinnerSize}px ${0.5 * spinnerSize}px 0px;
  animation: ${spinnerAnimation} 1s linear infinite;
  z-index: 99;
`;
