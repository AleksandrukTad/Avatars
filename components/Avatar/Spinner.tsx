import React from "react";

import { Circle, Spinner } from "./Spinner.style";

interface ILoadingProps extends React.Props<any> {
  loading: boolean;
}

const Loading: React.FunctionComponent<ILoadingProps> = ({ loading }) => {
  return (
    <Spinner loading={loading}>
      <Circle cx="20" cy="20" r="18" />
    </Spinner>
  );
};

export default Loading;
