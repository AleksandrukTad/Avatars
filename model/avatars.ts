import { Action, action } from "easy-peasy";

export interface IAvatar {
  index: number;
  name: string;
  styles: {
    borderColor: string;
    overlay?: string;
  };
}

export interface IAvatarsModel {
  active: IAvatar;
  items: IAvatar[];
  updateActive: Action<IAvatarsModel, IAvatar>;
}

const avatarsItems: IAvatar[] = [
  {
    index: 0,
    name: "Available",
    styles: {
      borderColor: "rgb(46, 204, 113)"
    }
  },
  {
    index: 1,
    name: "Not Read",
    styles: {
      borderColor: "rgb(251, 197, 49)",
      overlay: "rgb(251, 197, 49)"
    }
  },
  {
    index: 2,
    name: "Unavailable",
    styles: {
      borderColor: "rgb(194, 54, 22)",
      overlay: "rgb(0, 0, 0)"
    }
  }
];

const avatars: IAvatarsModel = {
  active: avatarsItems[0],
  items: avatarsItems,
  updateActive: action((state, newActive: IAvatar) => {
    state.active = newActive;
  })
};

export default avatars;
