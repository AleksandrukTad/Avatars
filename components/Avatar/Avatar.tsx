import React, { Component } from "react";

import Spinner from "./Spinner";

import Button from "../ui/Button/Button";

import { Circle, Text } from "./Avatar.style";

interface IProps extends React.Props<any> {
  active?: false | true;
  index?: number;
  styles: {
    borderColor: string;
    overlay?: string;
  };
  onEnter(bool: boolean): void;
  selectAvatar(index: number | undefined): void;
}

interface IState {
  extended: boolean;
  loading: boolean;
}

class Avatar extends Component<IProps, IState> {
  public static defaultProps = {
    active: false,
    index: 0,
    onEnter() {
      return null;
    },
    selectAvatar() {
      return null;
    },
    styles: {
      overlay: ""
    }
  };

  public state = {
    extended: false,
    loading: false
  };

  public selectAvatar = async () => {
    if (this.props.active) {
      return null;
    }

    await this.setState({ loading: true });

    setTimeout(async () => {
      await this.setState({ loading: false, extended: false });

      this.props.selectAvatar(this.props.index);
    }, 1000);
  };

  public onEnter = (extended: boolean): any => {
    this.setState({ extended });
  };

  public render() {
    const { onEnter, active, styles } = this.props;

    return (
      <Button
        onClick={this.selectAvatar}
        onMouseEnter={() => (active ? onEnter(true) : this.onEnter(true))}
        onMouseLeave={() => (active ? null : this.onEnter(false))}
      >
        <Spinner loading={this.state.loading} />

        <Circle
          active={!!active}
          // Avatar extends when you are hovering and when always when its loading
          extended={this.state.loading ? true : this.state.extended}
          loading={this.state.loading}
          {...styles}
        >
          <Text color="green"> T </Text>
          <Text color="blue"> A </Text>
        </Circle>
      </Button>
    );
  }
}

export default Avatar;
