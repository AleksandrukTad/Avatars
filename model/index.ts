import avatars, { IAvatarsModel } from "./avatars";

export interface IStoreModel {
  avatars: IAvatarsModel;
}

const model: IStoreModel = {
  avatars
};

export default model;
