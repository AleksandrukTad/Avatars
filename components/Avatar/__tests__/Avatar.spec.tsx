import "jest-dom/extend-expect";
import "jest-styled-components";

import { cleanup, render } from "@testing-library/react";
import React from "react";
// this adds custom jest matchers from jest-dom

import Avatar from "../Avatar";

import { IAvatar } from "../../../model/avatars";

const avatar: IAvatar = {
  index: 0,
  name: "Available",
  styles: {
    borderColor: "rgb(46, 204, 113)"
  }
};

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

test("Render Avatar", async () => {
  const { asFragment } = render(<Avatar {...avatar} />);
  // you can also get a `DocumentFragment`, which is useful if you want to compare nodes across renders
  expect(asFragment()).toMatchSnapshot();
});
