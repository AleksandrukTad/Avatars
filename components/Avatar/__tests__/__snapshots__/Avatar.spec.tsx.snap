// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Avatar 1`] = `
<DocumentFragment>
  .c1 {
  display: none;
  position: absolute;
  width: 40px;
  height: 40px;
  -webkit-transform: scale(1.5);
  -ms-transform: scale(1.5);
  transform: scale(1.5);
  z-index: 99999;
}

.c2 {
  fill: transparent;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  stroke-width: 4;
  stroke-linecap: round;
  stroke-dasharray: 125.60000000000001;
  -webkit-transform-origin: 20px 20px 0px;
  -ms-transform-origin: 20px 20px 0px;
  transform-origin: 20px 20px 0px;
  -webkit-animation: kRGcnj 1s linear infinite;
  animation: kRGcnj 1s linear infinite;
  z-index: 99;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 0px solid transparent;
  border-radius: 5px;
  cursor: pointer;
  line-height: 1.4;
  text-transform: uppercase;
  padding: 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 40px;
  width: 40px;
  border-radius: 50%;
  background-color: rgb(135,206,250);
  -webkit-transition: 0.25s ease-in-out;
  transition: 0.25s ease-in-out;
}

.c3::after {
  position: absolute;
  content: "";
  height: 40px;
  width: 40px;
  border-radius: 50%;
  background-color: transparent;
  border: solid 5px rgb(46,204,113);
}

.c3::before {
  position: absolute;
  content: "";
  width: 100%;
  height: 100%;
  border-radius: 50%;
  opacity: .5;
}

.c4 {
  color: green;
  margin: 0;
  padding: 0;
  font-weight: bold;
}

.c5 {
  color: blue;
  margin: 0;
  padding: 0;
  font-weight: bold;
}

<button
    class="c0"
    color=""
    height=""
    width=""
  >
    <svg
      class="c1"
    >
      <circle
        class="c2"
        cx="20"
        cy="20"
        r="18"
      />
    </svg>
    <div
      class="c3"
    >
      <p
        class="c4"
        color="green"
      >
        T
      </p>
      <p
        class="c5"
        color="blue"
      >
        A
      </p>
    </div>
  </button>
</DocumentFragment>
`;
